Dynamically load fdk-aac and FFmpeg at runtime.

Steveice10 opened the pull request:
Supercedes https://github.com/citra-emu/citra/pull/6568 as enough changes were made to warrant a fresh PR and review.

Introduces a cross-platform library loading utility, as well as dynamic loading of both fdk-aac and FFmpeg. For AAC decoding, when no OS decoders are enabled and neither are available, we fall back to the null decoder backend. For video dumping, when FFmpeg is not available we prompt the user to install a compatible version of FFmpeg, informing them which version of FFmpeg Citra is compatible with.

This has the following implications for users:
* On Linux, AppImage can now perform AAC decoding when the required libraries are installed on the system, using either fdk-aac or FFmpeg. Flatpak should always be able to decode AAC as the runtimes come with the required libraries.
* On all platforms, users will need to provide an FFmpeg install in order to use the video dumper. On most systems this should be trivial using a package manager. On Windows users will need to download FFmpeg from the official website, if they don't use one of the Windows package managers that are available.

Note that due to FFmpeg's design we can only reasonably create builds of Citra compatible with one major version of FFmpeg. I experimented with expanding to multiple versions but, for example, FFmpeg 6.x moved an important field in the AVStream struct that makes it incompatible with FFmpeg 5.1 and below.
* For Windows and macOS we can easily just target the latest version by keeping our bundled headers up to date and have people install that.
* For Flatpak the runtime is fixed so it won't be an issue.
* AppImage users will need to find and install the correct version, which should be 4.x based on the AppImage build environment. Worst case people can always compile against the FFmpeg installed on their system.

For the bundled headers I opted to include them directly into the repository instead of using submodules as FFmpeg's repository is ~250MB to clone, which is a lot for a few headers. I've trimmed them down to the bare minimum we need without modifying any of the headers.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/citra-emu/citra/6570)
<!-- Reviewable:end -->

FearlessTobi commented:
Please rebase.

Steveice10 commented:
> Please rebase.

Done

Steveice10 commented:
Conflicts with https://github.com/citra-emu/citra/pull/6565

Steveice10 commented:
Added an FFmpeg installation guide for the video dumper.

SachinVin commented:
Installation guide for windows seems complicated for the average user

Steveice10 commented:
> Installation guide for windows seems complicated for the average user
> 
> Edit: Yeah, no it seems like an absolute pain to get the right builds 
> 
> 

* Can you elaborate? You just need to download a zip and copy some DLL files.
* Do you have an alternative?

SachinVin commented:
> Can you elaborate? You just need to download a zip and copy some DLL files.

So, you go to the ffmpeg website, the first link is from gyan.dev, 
and if you somehow manage to navigate through their 2000's style webpage and download a build, these builds come with a static ffmpeg exe without any dlls.
And if you are adventurous and selected BtbN, the releases page by default only shows you `ffmpeg-master-latest-win64-gpl{-shared}.zip` which also don't come with any dlls; unless! you manage to find the expand button, and now you have to sift through 50+ releases, out of which there are 4 of them for windows with "n6.0..." in the name.
And then once you have downloaded the right version, you have to figure out where citra installed itself (I know you can open it from the menu, but it is not obvious for most people) and copy the dlls.

> Do you have an alternative?

1. Add a direct link to the right build
2. add a GUI to have the user select the ffmpeg folder or the ffmpeg dlls/exe (I'd personally love if this somehow happened automagically)

Steveice10 commented:
> So, you go to the ffmpeg website, the first link is from gyan.dev, 
> 
> and if you somehow manage to navigate through their 2000's style webpage and download a build, these builds come with a static ffmpeg exe without any dlls.

You need to download the archive marked with “shared” in the name, this is mentioned in the guide.

> 1. Add a direct link to the right build
> 
> 2. add a GUI to have the user select the ffmpeg folder or the ffmpeg dlls/exe (I'd personally love if this somehow happened automagically)

Seems reasonable enough, I’ll take a look at doing that. Someone will need to update the link every time the version changes though.

SachinVin commented:
gyan.dev doesn't have any archives named with the "n6.0..." convention, they call them `ffmpeg-release-full-shared.7z`, also they should get a better website

Steveice10 commented:
I have updated the guide to point to the download link and clarify some points: https://github.com/citra-emu/citra/wiki/Installing-FFmpeg-for-the-Video-Dumper

> Also Edge doesn't seem to like the downloads from gyan.dev and will complain about SmartScreen stuff

There’s nothing I can do about that, this is a build source recommended by FFmpeg themselves and the other source does not package release builds, only nightlies. If Edge has an issue with them they should fix their detection.

SachinVin commented:
> There’s nothing I can do about that, this is a build source recommended by FFmpeg themselves and the other source does not package release builds, only nightlies. If Edge has an issue with them they should fix their detection.

I know, just something to look out for

Steveice10 commented:
> I know, just something to look out for

I’m not an edge user myself so if you know how to get around that issue could you add it as a note to the wiki page?

Steveice10 commented:
Added the proposed utility to install FFmpeg into Citra, and updated the guide.

SachinVin commented:
Thanks, I'll take a look later today

SachinVin commented on src/citra_qt/main.cpp:
```diff
@@ -2182,7 +2159,92 @@ void GMainWindow::OnCaptureScreenshot() {
+    if (filename.empty()) {
+        return;
+    }
+    auto bin_dir = filename + DIR_SEP + "bin";
```
maybe it is just me, but if I selected the bin directory instead of the ffmpeg directory. it will fail and saying it couldn't find the dlls

Steveice10 replied:
Should work with both now.

SachinVin commented:
Because the installer/updater is kinda shit, it is probably going to delete the dlls when citra updates

Steveice10 commented:
> Because the installer/updater is kinda shit, it is probably going to delete the dlls when citra updates

Updating on my system with the Qt installer did not delete the DLLs.

SachinVin commented:
Alright, Sounds good.

Steveice10 commented:
Per discussion elsewhere, I've moved the library headers to a separate submodule and rebased. The submodule repo will need to be forked into the Citra org.

GPUCode commented on src/common/dynamic_library/dynamic_library.h:
```diff
@@ -0,0 +1,119 @@
+
+namespace DynamicLibrary {
+
+#if defined(_WIN32)
```
Nit, most of these could be moved into a cpp file. Also can be constexpr std::string_view or better make them into a struct

GPUCode commented on src/common/dynamic_library/dynamic_library.h:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    static std::string GetLibraryName(const std::string& name, u32 version) {
+        std::stringstream lib_name;
```
I think the way yuzu does this is easier to understand https://github.com/yuzu-emu/yuzu/blob/4112031c81a7bd3d80c541a85e8e7dc717820aeb/src/common/dynamic_library.cpp#L51

GPUCode commented on src/common/dynamic_library/dynamic_library.h:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    template <typename T>
+    struct _symbol_type_helper {
```
Why is this helper needed?

Steveice10 replied:
I remember having some issue with the generics but it doesn't seem to be a problem with the current iteration of the code, so removed.

GPUCode commented on src/common/dynamic_library/dynamic_library.h:
```diff
@@ -0,0 +1,119 @@
+
+class DynamicLibrary {
+public:
+    explicit DynamicLibrary(const std::string& name, u32 version = 0) {
```
```suggestion
    explicit DynamicLibrary(std::string_view name, u32 version = 0) {
```

theofficialgman commented:
breaks building from source, refer to the log here https://github.com/cobalt2727/L4T-Megascript/issues/139#issue-1762437281

if you need to use a specific version of ffmpeg, you should include it as a submodule and bundle it in the application. any other method is incorrect and unstable.

Steveice10 commented:
> if you need to use a specific version of ffmpeg, you should include it as a submodule and bundle it in the application. any other method is incorrect and unstable.

We support FFmpeg 4.2 and up and do bundle headers that we use when the user is missing system headers.

cobalt2727 commented:
> We support FFmpeg 4.2 and up and do bundle headers that we use when the user is missing system headers.

Seems like that might not always be the case right now, unless I'm misunderstanding something - 18.04 is probably an edge case that was missed during testing of this PR:

https://github.com/citra-emu/citra/issues/6624

theofficialgman commented:
CI is running now. This will take a few hours but we will see what passes and what doesn't https://github.com/cobalt2727/L4T-Megascript/actions/runs/5305595133

Steveice10 commented:
This might help: https://github.com/citra-emu/ext-library-headers/pull/1

theofficialgman commented:
CI confirms the issue on bionic https://github.com/cobalt2727/L4T-Megascript/actions/runs/5305595133/jobs/9602657474

cobalt2727 commented:
> CI confirms the issue on bionic https://github.com/cobalt2727/L4T-Megascript/actions/runs/5305595133/jobs/9602657474

Re-running now. Fingers crossed.
